File: app/src/androidTest/java/at/priv/graf/zazentimer/screens/SessionEditPage.java
package at.priv.graf.zazentimer.screens;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.Espresso.pressBack;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;

import android.os.SystemClock;

import at.priv.graf.zazentimer.R;

public class SessionEditPage extends BasePage {

    public SessionEditPage verifyEditSessionScreen() {
        checkElementIsDisplayed(R.id.text_sitzung_name);
        checkElementIsDisplayed(R.id.but_new_section);
        return this;
    }

    public SessionEditPage setSessionName(String name) {
        onView(withId(R.id.text_sitzung_name)).perform(typeText(name));
        return this;
    }

    public SessionEditPage setSessionDescription(String desc) {
        onView(withId(R.id.text_sitzung_beschreibung)).perform(typeText(desc));
        return this;
    }

    public SectionEditPage clickAddSection() {
        clickOnView(R.id.but_new_section);
        return new SectionEditPage();
    }

    public SessionEditPage clickSectionAtPosition(int pos) {
        onView(withId(R.id.list))
                .perform(actionOnItemAtPosition(pos, click()));
        return this;
    }

    public SessionEditPage verifySectionCount(int count) {
        assertRecyclerViewItemCount(R.id.list, count);
        return this;
    }

    public MainPage goBack() {
        pressBack();
        SystemClock.sleep(800);
        return new MainPage();
    }
}

File: app/src/androidTest/kotlin/at/priv/graf/zazentimer/screens/SessionEditPage.kt
package at.priv.graf.zazentimer.screens

import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.typeText
import androidx.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition
import androidx.test.espresso.matcher.ViewMatchers.withId
import at.priv.graf.zazentimer.R
import at.priv.graf.zazentimer.utils.ScreenRobot

class SessionEditPage {
    private val robot = ScreenRobot()

    fun verifyEditSessionScreen(): SessionEditPage {
        Thread.sleep(1000)
        robot.checkElementIsDisplayed(R.id.text_session_name)
        robot.checkElementIsDisplayed(R.id.section_list)
        return this
    }

    fun setSessionName(name: String): SessionEditPage {
        onView(withId(R.id.text_session_name)).perform(typeText(name))
        return this
    }

    fun setSessionDescription(desc: String): SessionEditPage {
        onView(withId(R.id.text_session_description)).perform(typeText(desc))
        return this
    }

    fun clickAddSection(): SectionEditPage {
        robot.clickToolbarOverflowItem(R.string.menu_add_section)
        Thread.sleep(1500)
        return SectionEditPage()
    }

    fun clickSectionAtPosition(pos: Int): SessionEditPage {
        robot.waitForRecyclerViewToBePopulated(R.id.section_list)
        for (i in 0 until 10) {
            try {
                onView(withId(R.id.section_list))
                    .perform(actionOnItemAtPosition<RecyclerView.ViewHolder>(pos, click()))
                return this
            } catch (e: Throwable) {
                Thread.sleep(500)
            }
        }
        throw AssertionError("Could not click section at position $pos")
    }

    fun verifySectionCount(count: Int): SessionEditPage {
        robot.assertRecyclerViewItemCount(R.id.section_list, count)
        return this
    }

    fun goBack(): MainPage {
        robot.pressBack()
        Thread.sleep(3000)
        return MainPage()
    }
}
